docs/middleware/callbacks/integration.md

```diff
@@ -0,0 +1,100 @@
+
+The callbacks middleware is a minimal and stateless implementation of the IBC middleware interface. It does not have a keeper, nor does it store any state. It simply routes IBC middleware messages to the appropriate callback function, which is implemented by the secondary application. Therefore, it doesn't need to be registered as a module, nor does it need to be added to the module manager. It only needs to be added to the IBC application stack.
+
+## Pre-requisite Readings
+
+* [IBC middleware development](../../ibc/middleware/develop.md) {prereq}
+* [IBC middleware integration](../../ibc/middleware/integration.md) {prereq}
+
+The callbacks middleware, as the name suggests, plays the role of an IBC middleware and as such must be configured by chain developers to route and handle IBC messages correctly.
+For Cosmos SDK chains this setup is done via the `app/app.go` file, where modules are constructed and configured in order to bootstrap the blockchain application.
+
+## Configuring an application stack with the callbacks middleware
+
+As mentioned in [IBC middleware development](../../ibc/middleware/develop.md) an application stack may be composed of many or no middlewares that nest a base application.
+These layers form the complete set of application logic that enable developers to build composable and flexible IBC application stacks.
+For example, an application stack may just be a single base application like `transfer`, however, the same application stack composed with `29-fee` and `callbacks` will nest the `transfer` base application twice by wrapping it with the Fee Middleware module and then callbacks middleware.
+
+The callbacks middleware also **requires** a secondary application that will receive the callbacks to implement the [`ContractKeeper`](https://github.com/cosmos/ibc-go/blob/main/modules/apps/callbacks/types/expected_keepers.go#L11-L64). Since the wasm module does not yet support the callbacks middleware, we will use the `mockContractKeeper` module in the examples below. You should replace this with your own module that implements `ContractKeeper`.
+
+### Transfer
+
+See below for an example of how to create an application stack using `transfer`, `29-fee`, and `callbacks`. Feel free to omit the `29-fee` middleware if you do not want to use it.
+The following `transferStack` is configured in `app/app.go` and added to the IBC `Router`.
+The in-line comments describe the execution flow of packets between the application stack and IBC core.
+
+```go
+// Create Transfer Stack
+// SendPacket, since it is originating from the application to core IBC:
+// transferKeeper.SendPacket -> callbacks.SendPacket -> fee.SendPacket -> channel.SendPacket
+
+// RecvPacket, message that originates from core IBC and goes down to app, the flow is the other way
+// channel.RecvPacket -> callbacks.OnRecvPacket -> fee.OnRecvPacket -> transfer.OnRecvPacket
```

Comment:
However, I think the way we currently wired it up makes it so that callbacks is called before fee in both cases. So, I don't think your description above is correct even with the new code snippet.

Reply:
Sorry I don't follow. callbacks should not be before fee in both cases of traversing the stack up and down. If callbacks was a module that was stateful (and had a version), then wiring the stack like this would break `GetAppVersion`. The version would be wrapped differently then it would be unwrapped. 

See example:
Stack: A, B, C, D
Upwards motion: A -> B -> C -> D
Downwards motion: D -> C -> B -> A

How this example wires:
Stack: A, B, C, D
Upwards motion: A -> C -> B -> D
Downwards motion: D -> C -> B -> A

Wrapping version ordering: D ( C ( B ( A ) ) ) )
Unwrapping version: D -> B -> C -> A, this will fail, as C should be unwrapped before B

The problem I foresee is that we should only have 1 keeper per module but that we may have many instances of modules. Each IBC app might have different stack orderings:
`transfer -> callbacks -> fee -> channel`
`ica -> callbacks -> fee -> custom module -> channel`

By having the ics4wrapper on the keeper, the above wouldn't be possible without some odd hacks

My suggestion is to make callbacks be the middle module in the example since the ics4wrapper is on the IBC module, thus the fee keeper can keep channel as its ics4wrapper. As the middle module, callbacks could be wired up to take different ics4wrappers in the transfer stack and the ica stack

Does that make sense?

Reply:
Yes, I understand now. I think we should open a new PR/issue for this as we should run all the tests again with the new wiring right?

Reply:
I've created an issue for this https://github.com/cosmos/ibc-go/issues/4444

Reply:
Yea lets do that. Happy to leave this as is for v7.3